common/views/components/IIIFViewer/parts/MainViewer.js

```diff
@@ -0,0 +1,244 @@
+  const rotation = matching ? matching.rotation : 0;
+  const imageType = getImageTypeForScrollVelocity(scrollVelocity);
+  return (
+    <div style={style}>
+      {scrollVelocity === 3 || isProgrammaticScroll ? (
+        <div style={{ display: 'flex', justifyContent: 'center' }}>
+          <LL lighten={true} />
+        </div>
+      ) : (
+        <>
+          <LL lighten={true} />
+          {!mainLoaded && urlTemplateThumbnail && (
+            <ThumbnailWrapper imageLoaded={thumbLoaded}>
+              <IIIFResponsiveImage
+                width={currentCanvas.width}
+                height={currentCanvas.height}
+                src={urlTemplateThumbnail({
+                  size: `${
+                    smallestWidthImageDimensions
+                      ? smallestWidthImageDimensions.width
+                      : '!100'
+                  },`,
+                })}
+                srcSet={''}
+                sizes={''}
+                alt={''}
+                isLazy={false}
+                lang={null}
+                onLoadHandler={() => {
+                  setThumbLoaded(true);
+                }}
+              />
+            </ThumbnailWrapper>
+          )}
+          {(imageType === 'main' || mainLoaded) && urlTemplateMain && (
+            <ImageViewer
+              id="item-page"
+              infoUrl={infoUrl}
+              width={currentCanvas.width}
+              height={currentCanvas.height}
+              alt={ocrText}
+              urlTemplate={urlTemplateMain}
+              setShowZoomed={setShowZoomed}
+              setZoomInfoUrl={setZoomInfoUrl}
+              rotation={rotation}
+              mainViewerRef={
+                mainViewerRef && mainViewerRef.current
+                  ? mainViewerRef.current
+                  : null
+              }
+              setActiveIndex={setActiveIndex}
+              index={index}
+              onLoadHandler={() => {
+                setMainLoaded(true);
+                setIsLoading(false);
+              }}
+            />
+          )}
+        </>
+      )}
+    </div>
+  );
+}, areEqual);
+
+type Props = {|
+  listHeight: number,
+  mainViewerRef: any,
+  setActiveIndex: number => void,
+  pageWidth: number,
+  canvases: [],
+  canvasIndex: number,
+  setShowZoomed: boolean => void,
+  setZoomInfoUrl: string => void,
+  rotatedImages: [],
+  setShowControls: boolean => void,
+  setIsLoading: boolean => void,
+|};
+
+const MainViewer = ({
+  listHeight,
+  mainViewerRef,
+  setActiveIndex,
+  pageWidth,
+  canvases,
+  canvasIndex,
+  setShowZoomed,
+  setZoomInfoUrl,
+  rotatedImages,
+  setShowControls,
+  setIsLoading,
+}: Props) => {
+  const [isProgrammaticScroll, setIsProgrammaticScroll] = useState(false);
+  const [newScrollOffset, setNewScrollOffset] = useState(0);
+  const [firstRender, setFirstRender] = useState(true);
+  const firstRenderRef = useRef(firstRender);
+  firstRenderRef.current = firstRender;
```

Comment:
I'm guessing this line is necessary, but I can't work out why (i.e. how is it any different in practice to the line above?).

Reply:
yes, it is needed. For similar reasons to discussed in this issue: https://github.com/facebook/react/issues/14010#issuecomment-433788147

Which I think explains it better than I can

Reply:
aaaah – got it.